ignite/metrics/multilabel_confusion_matrix.py

```diff
@@ -0,0 +1,112 @@
+import numbers
+from typing import Callable, Optional, Sequence, Tuple, Union
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+from ignite.metrics.metrics_lambda import MetricsLambda
+
+__all__ = [
+    "MultiLabelConfusionMatrix",
+]
+
+
+class MultiLabelConfusionMatrix(Metric):
+    def __init__(
+        self,
+        num_classes: int,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+        normalized: bool = False,
+    ):
+        if num_classes <= 1:
+            raise ValueError("Argument num_classes needs to be > 1")
+
+        self.num_classes = num_classes
+        self._num_examples = 0
+        self.normalized = normalized
+        super(MultiLabelConfusionMatrix, self).__init__(output_transform=output_transform, device=device)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self.confusion_matrix = torch.zeros(self.num_classes, 2, 2, dtype=torch.int64, device=self._device)
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        self._check_input(output)
+        y_pred, y = output[0].detach(), output[1].detach()
+
+        self._num_examples += y.shape[0]
+        y_reshaped = y.swapaxes(0, 1).reshape(self.num_classes, -1)
+        y_pred_reshaped = y_pred.swapaxes(0, 1).reshape(self.num_classes, -1)
+        y_total = torch.unsqueeze(torch.count_nonzero(y_reshaped, axis=1), 0)
+        y_pred_total = torch.unsqueeze(torch.count_nonzero(y_pred_reshaped, axis=1), 0)
+        tp = torch.unsqueeze(torch.count_nonzero(torch.mul(y_reshaped, y_pred_reshaped), axis=1), 0)
+        fp = y_pred_total - tp
+        fn = y_total - tp
+        tn = y_reshaped.shape[1] - tp - fp - fn
+
+        self.confusion_matrix += torch.cat([tn, fp, fn, tp], dim=0).T.reshape(-1, 2, 2)
+
+    @sync_all_reduce("confusion_matrix", "_num_examples")
+    def compute(self) -> torch.Tensor:
+        if self._num_examples == 0:
+            raise NotComputableError("Confusion matrix must have at least one example before it can be computed.")
+
+        if self.normalized:
+            conf = self.confusion_matrix.type(torch.float64)
+            sums = conf.sum(axis=2).sum(axis=1)
+            return (conf.reshape(conf.shape[0], 4) / sums[:, None]).reshape(conf.shape)
```

Comment:
Shouldn't we add eps=1e-15 to sums if it is zero or we can admit `inf` ?

Reply:
By definition, all the entries of ```sums``` will be equal to the number of configurations for each class. For instance in the case of 2D ```y_pred``` and ```y```, the entries of ```sums``` will all be equal to the batch size.